File: src/DataBaseController.java
import org.neo4j.driver.Config;
import org.neo4j.driver.types.Node;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;

public class DataBaseController {

    private String URI= System.getenv("NEO4J_URI");
    private String USER = System.getenv("NEO4J_USERNAME");
    private String PASSWORD = System.getenv("NEO4J_PASSWORD");
    private User currentUser;
    private ArrayList<Game> filteredGames = new ArrayList<>();
    private ArrayList<Game> recommendedgames = new ArrayList<>();

    // Getters for user-related data
    // This will only have data after the user has logged in or has previously selected an action
    // Such as get recommendations for example
    public ArrayList<Game> getUserPlayedGames(){
        return currentUser.getPlayedGames();
    }

    public ArrayList<Game> getUserFavoriteGames(){
        return currentUser.getFavoriteGames();
    }

    public ArrayList<Game> getFilteredGames() {
        return filteredGames;
    }

    public ArrayList<Game> getRecommendedgames() {
        return recommendedgames;
    }

    // Check if an account exists in the database
    public boolean accountExists(String name, String password){
        try (var app = new DataBaseDriver(URI, USER, PASSWORD, Config.defaultConfig())) {
            if(!app.userExists(name, password)){
                return true;
            }
        }
        return false;
    }

    // Validate login credentials and set the current user and filtered games
    // Once the user has logged in, the program automatically gets the user from the database and filters
    // The games according to the user preferences and age
    public void validLogin(String name, String password){
        try (var app = new DataBaseDriver(URI, USER, PASSWORD, Config.defaultConfig())) {
            if(accountExists(name, password)){
                setCurrentUserFromDataBase(name, password);
                setFilteredGamesFromDataBase();
                filteredGames = getFilteredGamesByESRB(filteredGames, currentUser);
            }
        }
    }

    // Set the current user based on the provided name and password
    public void setCurrentUserFromDataBase(String name, String password){
        try (var app = new DataBaseDriver(URI, USER, PASSWORD, Config.defaultConfig())) {
            if(app.userExists(name, password)){
                currentUser = app.mapUser(app.findUserNode(name, password));
            }
        }
    }

    // Set the filtered games based on the current user
    public void setFilteredGamesFromDataBase(){
        try (var app = new DataBaseDriver(URI, USER, PASSWORD, Config.defaultConfig())) {
            if(currentUser != null){
                filteredGames = app.getCompatibleGames(currentUser);
            }
        }
    }

    // Validate sign-in credentials, create a new account, and set the current user and filtered games
    // If the user does not exist a new user is created and its favorite games are saved
    public void validSignIn(String name, String password, int age, boolean preferNintendo, boolean preferPC,
                            boolean preferMobile, boolean preferXbox, boolean preferPlayStation, boolean preferMulti, ArrayList<String> playedGamesNames){
        try (var app = new DataBaseDriver(URI, USER, PASSWORD, Config.defaultConfig())) {
            if(!accountExists(name, password)){
                app.crearNodoPersona(name, age, password, preferNintendo, preferPC, preferMobile, preferXbox, preferPlayStation, preferMulti);
                for (String gameName: playedGamesNames){
                    app.crearRelacionPersonaJuego(name,gameName);
                }
                validLogin(name, password);
            }
        }
    }

    // Filter games based on ESRB rating and user's age
    public ArrayList<Game> getFilteredGamesByESRB(ArrayList<Game> games, User user) {
        int userAge = user.getUserAge();
        ArrayList<Game> permittedGames = new ArrayList<>();

        for (Game currentGame : games) {
            String gameRating = currentGame.getEsrbRating();

            if (userAge < 17 && userAge >= 13) {
                if (!gameRating.equals("M")) {
                    permittedGames.add(currentGame);
                }
            } else if (userAge < 13 && userAge >= 10) {
                if (!gameRating.equals("M") && !gameRating.equals("T")) {
                    permittedGames.add(currentGame);
                }
            } else if (userAge < 10) {
                if (!gameRating.equals("M") && !gameRating.equals("T") && !gameRating.equals("E10+")) {
                    permittedGames.add(currentGame);
                }
            } else {
                // For users 17 and above, all games are permitted
                permittedGames.add(currentGame);
            }
        }
        return permittedGames;
    }

    // Filter games based on categories
    //When a user tries to do a filtered search by selecting categories
    public ArrayList<Game> getGamesFilteredByCategory(ArrayList<String> Categories){
        ArrayList<Game> newFilteredGames = new ArrayList<>();

        for (Game game: filteredGames){
            if(Categories.contains(game.getCategory1()) || Categories.contains(game.getCategory2()) || Categories.contains(game.getCategory3())){
                filteredGames.add(game);
            }
        }
        return newFilteredGames;
    }

    // Recommend games based on user's played and favorite games
    // The games are stored in the variable of the class recommendedgames
    public ArrayList<Game> recommendGames() {
        ArrayList<Game> playedGames = currentUser.getPlayedGames();
        ArrayList<Game> favoriteGames = currentUser.getFavoriteGames();
        ArrayList<Game> allGames = filteredGames;

        // Add favorite games to the played games list
        for (Game game : favoriteGames) {
            if (!playedGames.contains(game)) {
                playedGames.add(game);
            }
        }

        int[][] similarityMatrix = new int[playedGames.size()][favoriteGames.size()];

        // Calculate similarity scores and populate the similarity matrix
        for (int i = 0; i < playedGames.size(); i++) {
            Game playedGame = playedGames.get(i);

            for (int j = 0; j < favoriteGames.size(); j++) {
                Game favoriteGame = favoriteGames.get(j);

                // Calculate similarity score for each game pair
                int similarityScore = calculateSimilarity(playedGame, favoriteGame);

                // Assign similarity score to the matrix
                similarityMatrix[i][j] = similarityScore;
            }
        }

        int numRecommendations = Math.min(5, allGames.size()); // Get up to 5 recommendations

        // Generate recommendations based on the similarity matrix
        for (int k = 0; k < numRecommendations; k++) {
            int maxSimilarity = Integer.MIN_VALUE;
            int maxSimilarityIndex = -1;

            for (int i = 0; i < playedGames.size(); i++) {
                for (int j = 0; j < favoriteGames.size(); j++) {
                    int similarity = similarityMatrix[i][j];
                    if (similarity > maxSimilarity) {
                        maxSimilarity = similarity;
                        maxSimilarityIndex = j;
                    }
                }
            }

            if (maxSimilarityIndex != -1) {
                Game recommendedGame = allGames.get(maxSimilarityIndex);
                recommendedgames.add(recommendedGame);
                // Set the similarity score to -1 to avoid selecting it again
                Arrays.fill(similarityMatrix[maxSimilarityIndex], -1);
            }
        }

        return recommendedgames;
    }

    // Calculate similarity between two games based on certain criteria
    private int calculateSimilarity(Game playedGame, Game game) {
        int similarity = 0;

        if (playedGame.getDuration().equals(game.getDuration())) {
            similarity += 5;
        }
        if (playedGame.getCategory1().equals(game.getCategory1())
                || playedGame.getCategory1().equals(game.getCategory2())
                || playedGame.getCategory1().equals(game.getCategory3())) {
            similarity += 10;
        }
        if (playedGame.getCategory2().equals(game.getCategory1())
                || playedGame.getCategory2().equals(game.getCategory2())
                || playedGame.getCategory2().equals(game.getCategory3())) {
            similarity += 10;
        }
        if (playedGame.getCategory3().equals(game.getCategory1())
                || playedGame.getCategory3().equals(game.getCategory2())
                || playedGame.getCategory3().equals(game.getCategory3())) {
            similarity += 10;
        }

        return similarity;
    }

    public ArrayList<String> getAllExistingCategories(){
        ArrayList<String> allCategories = new ArrayList<>();

        try (var app = new DataBaseDriver(URI, USER, PASSWORD, Config.defaultConfig())) {
            ArrayList<Game> allGames = app.getAllGames();

            for ( Game game : allGames){
                if (!allCategories.contains(game.getCategory1())){
                    allCategories.add(game.getCategory1());
                }
                if (!allCategories.contains(game.getCategory2())){
                    allCategories.add(game.getCategory2());
                }
                if (!allCategories.contains(game.getCategory3())){
                    allCategories.add(game.getCategory3());
                }
            }
        }
        return allCategories;
    }
}

File: apps/DataBaseController.py
import os
from dotenv import load_dotenv
from neo4j import GraphDatabase
from apps.DataBaseDriver import *

class DataBaseController:
    def __init__(self):
        load_dotenv('./.env')
        self.URI = os.getenv("NEO4J_URI")
        self.USER = os.getenv("NEO4J_USERNAME")
        self.PASSWORD = os.getenv("NEO4J_PASSWORD")
        self.currentUser = None
        self.filteredGames = []
        self.recommendedGames = []

    def get_user_played_games(self):
        return self.currentUser.played_games

    def get_user_favorite_games(self):
        return self.currentUser.favorite_games

    def account_exists(self, name, password):
        with DataBaseDriver(self.URI, self.USER, self.PASSWORD, Config.default_config()) as app:
            if not app.user_exists(name, password):
                return True
        return False

    def valid_login(self, name, password):
        with DataBaseDriver(self.URI, self.USER, self.PASSWORD, Config.default_config()) as app:
            if self.account_exists(name, password):
                self.set_current_user_from_database(name, password)
                self.set_filtered_games_from_database()
                self.filteredGames = self.get_filtered_games_by_esrb(self.filteredGames, self.currentUser)

    def set_current_user_from_database(self, name, password):
        with DataBaseDriver(self.URI, self.USER, self.PASSWORD, Config.default_config()) as app:
            if app.user_exists(name, password):
                user_node = app.find_user_node(name, password)
                self.currentUser = User(
                    user_node.name,
                    user_node.password,
                    user_node.age,
                    user_node.prefers_nintendo,
                    user_node.prefers_pc,
                    user_node.prefers_mobile,
                    user_node.prefers_xbox,
                    user_node.prefers_playstation,
                    user_node.prefers_multiplayer,
                    user_node.played_games,
                    user_node.favorite_games
                )

    def set_filtered_games_from_database(self):
        with DataBaseDriver(self.URI, self.USER, self.PASSWORD, Config.default_config()) as app:
            if self.currentUser is not None:
                self.filteredGames = app.get_compatible_games(self.currentUser)

    def valid_sign_in(self, name, password, age, prefer_nintendo, prefer_pc, prefer_mobile, prefer_xbox,
                      prefer_playstation, prefer_multi):
        with DataBaseDriver(self.URI, self.USER, self.PASSWORD, Config.default_config()) as app:
            if not self.account_exists(name, password):
                app.create_persona_node(name, age, password, prefer_nintendo, prefer_pc, prefer_mobile, prefer_xbox,
                                        prefer_playstation, prefer_multi)
                self.valid_login(name, password)

    def get_filtered_games_by_esrb(self, games, user):
        user_age = user.user_age
        permitted_games = []

        for current_game in games:
            game_rating = current_game.esrb_rating

            if user_age < 17 and user_age >= 13:
                if game_rating != "M":
                    permitted_games.append(current_game)
            elif user_age < 13 and user_age >= 10:
                if game_rating != "M" and game_rating != "T":
                    permitted_games.append(current_game)
            elif user_age < 10:
                if game_rating != "M" and game_rating != "T" and game_rating != "E10+":
                    permitted_games.append(current_game)
            else:
                # For users 17 and above, all games are permitted
                permitted_games.append(current_game)

        return permitted_games

    def get_games_filtered_by_category(self, categories):
        new_filtered_games = []

        for game in self.filteredGames:
            if game.categories[0] in categories or game.categories[1] in categories or game.categories[2] in categories:
                new_filtered_games.append(game)

        return new_filtered_games